mypy/checkexpr.py

```diff
@@ -5095,7 +5095,11 @@ def visit_yield_from_expr(self, e: YieldFromExpr, allow_none_return: bool = Fals
 
         # Determine the type of the entire yield from expression.
         iter_type = get_proper_type(iter_type)
-        if isinstance(iter_type, Instance) and iter_type.type.fullname == "typing.Generator":
+        if (
+            isinstance(iter_type, Instance)
+            and iter_type.type.fullname == "typing.Generator"
+            or isinstance(iter_type, UnionType)
```

Comment:
10,000ft / 30,000m high observation:

This reads unbalanced to me: one case (`typing.Generator`) is very specific to generators, while the other case is broad (a union of anything).

Maybe the condition is logic that should move into `get_generator_return_type`?

Reply:
Thanks for the observation, I've committed something that uses the suggestion from @hauntsaninja and fixed the failing test.